### PR TITLE
Added additional Linq extension methods for First, Select, Take, Skip…

### PR DESCRIPTION
**Pull request by ppittle**
… and Where

### CODE REVIEW

**Comment by psibernetic** [comment]
Good shit.

File: AsyncEnumerable.nuspec [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
     <metadata>
         <id>AsyncEnumerator</id>
-        <version>1.2.0</version>
+        <version>1.2.1</version>
```
Comment (ppittle) [inline AsyncEnumerable.nuspec]: Was just trying to be helpful :).  All commnets have been resolved and rebased.  Let me know if you still want me to fully revert this commit

File: IAsyncEnumerableExtensions.cs [hunk 2]
```diff
@@ -1,15 +1,228 @@
...
+            this IAsyncEnumerable<TSource> source,
+            Func<TSource, bool> predicate,
+            CancellationToken token = default(CancellationToken))
+        {
+            if (null == source)
+                throw new ArgumentNullException(nameof(source));
+            if (null == predicate)
+                throw new ArgumentNullException(nameof(predicate));
+
+            using (var enumerator = await source.GetAsyncEnumeratorAsync(token))
```
Comment (ppittle) [inline IAsyncEnumerableExtensions.cs]: fixed

File: IAsyncEnumerableExtensions.cs [hunk 3]
```diff
@@ -1,15 +1,228 @@
...
+        {
+            if (null == source)
+                throw new ArgumentNullException(nameof(source));
+            if (null == selector)
+                throw new ArgumentNullException(nameof(selector));
+
+            return new AsyncEnumerable<TResult>(
+                yield =>
+                    source.ForEachAsync(
+                        async item => await yield.ReturnAsync(selector(item)).ConfigureAwait(false),
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: this line does not require ```async``` and ```await```

File: IAsyncEnumerableExtensions.cs [hunk 4]
```diff
@@ -1,15 +1,228 @@
...
+        {
+            if (null == source)
+                throw new ArgumentNullException(nameof(source));
+            if (null == selector)
+                throw new ArgumentNullException(nameof(selector));
+
+            return new AsyncEnumerable<TResult>(
+                yield =>
+                    source.ForEachAsync(
+                        async (item, index) => await yield.ReturnAsync(selector(item, index)).ConfigureAwait(false),
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: this line does not require ```async```, ```await```, and ```ConfigureAwait```

File: IAsyncEnumerableExtensions.cs [hunk 5]
```diff
@@ -1,15 +1,228 @@
...
+            this IAsyncEnumerable<TSource> source,
+            int count,
+            bool oneTimeUse = false,
+            CancellationToken token = default(CancellationToken))
+        {
+            if (null == source)
+                throw new ArgumentNullException(nameof(source));
+
+            if (count <= 0)
+                return new AsyncEnumerable<TSource>(yield => Task.FromResult(Enumerable.Empty<TSource>()));
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: ```AsyncEnumerable.Empty<TSource>```

File: IAsyncEnumerableExtensions.cs [hunk 6]
```diff
@@ -1,15 +1,228 @@
...
+
+            return new AsyncEnumerable<TSource>(
+                async yield =>
+                {
+                    using (var enumerator = await source.GetAsyncEnumeratorAsync(token).ConfigureAwait(false))
+                        while (count > 0)
+                        {
+                            if (await enumerator.MoveNextAsync(token))
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: ```ConfigureAwait(false)```

File: IAsyncEnumerableExtensions.cs [hunk 7]
```diff
@@ -38,5 +255,138 @@ public static class IAsyncEnumerableExtensions
...
+                throw new ArgumentNullException(nameof(source));
+            if (null == predicate)
+                throw new ArgumentNullException(nameof(predicate));
+
+            return new AsyncEnumerable<TSource>(
+                yield =>
+                    source.ForEachAsync(async item =>
+                    {
+                        if (predicate(item))
+                            await yield.ReturnAsync(item).ConfigureAwait(false);
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: No need for ```async```:
     if (predicate(item))
         return yield.ReturnAsync(item);
    else
        return TaskEx.Completed;

p.s. please add ```TaskEx```

Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: Or just use enumerator instead of ```ForEachAsync```

File: IAsyncEnumerableExtensions.cs [hunk 8]
```diff
@@ -38,5 +255,138 @@ public static class IAsyncEnumerableExtensions
...
+                throw new ArgumentNullException(nameof(source));
+            if (null == predicate)
+                throw new ArgumentNullException(nameof(predicate));
+
+            return new AsyncEnumerable<TSource>(
+                yield =>
+                    source.ForEachAsync(async (item, index) =>
+                    {
+                        if (predicate(item, index))
+                            await yield.ReturnAsync(item).ConfigureAwait(false);
```
Comment (tyrotoxin) [inline IAsyncEnumerableExtensions.cs]: same comment as for another overloaded ```WhereAsync```

File: IAsyncEnumerableExtensions.cs (hunk 3 above)
Comment (ppittle) [inline IAsyncEnumerableExtensions.cs]: done

File: IAsyncEnumerableExtensions.cs (hunk 4 above)
Comment (ppittle) [inline IAsyncEnumerableExtensions.cs]: nice catch!  fixed

**Pull request merged**